src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.WriteCore.cs

```diff
@@ -17,6 +17,14 @@ public partial class JsonConverter<T>
                 ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
             }
 
+            // Root object is a boxed value type.
+            // Since value types are ignored when detecting cycles
+            // We need to push it before it gets unboxed.
+            if (value != null && IsValueType && JsonSerializer.IsIgnoreCyclesEnabled(options))
```

Comment:
It is unfortunate that we need to add code in this speed-critical method. Can we at least combine the previous `if` statement that also checks for `IsValueType`?

Reply:
Done. 

However I was wondering why do we need to unbox the root object here. When a struct is boxed in `object` couldn't be possible to just call `JsonConverter<object>.WriteCore()` and expect that the unboxing happens on the `TryWrite` method, similar to how polymorphic properties do?